KitePrintSDK/src/main/java/ly/kite/print/ProductLoader.java

```diff
@@ -0,0 +1,892 @@
+          {
+          // Ignore
+          }
+
+
+        // Create the product and display it
+
+        Product product = new Product( productId, productCode, productName, productType, labelColour, userJourneyType, imagesPerPage )
+                .setCost( cost )
+                .setShippingCosts( shippingCosts )
+                .setImageURLs( heroImageURL, imageURLList )
+                .setLabelColour( labelColour )
+                .setMask( maskURL, maskBleed )
+                .setSize( size );
+
+        Log.i( LOG_TAG, "-- Found product --" );
+        Log.i( LOG_TAG, product.toLogString() );
+
+
+
+        // Only display products for which we have a defined user journey
+        if ( ! UserJourneyCoordinator.getInstance().isSupported( userJourneyType ) )
+          {
+          Log.i( LOG_TAG, "-- Product discarded: no user journey --" );
+
+          continue next_product;
+          }
+
+
+        // See if we already have the product group. If not - create it now.
+
+        ProductGroup productGroup = productGroupTable.get( groupLabel );
+
+        if ( productGroup == null )
+          {
+          productGroup = new ProductGroup( groupLabel, labelColour, groupImageURL );
+
+          productGroupList.add( productGroup );
+          productGroupTable.put( groupLabel, productGroup );
+          }
+
+
+        // Add the product to its group
+        productGroup.add( product );
+
+        // Add the product to the product id / product table
+        productTable.put( productId, product );
+        }
+      catch ( Exception exception )
+        {
+        Log.e( LOG_TAG, "Unable to parse JSON product: " + productJSONObject, exception );
+
+        // Ignore individual errors - try and getCost as many products as possible
+        }
+      }
+
+
+    return ( new Pair<>( productGroupList, productTable ) );
+    }
+
+
+  ////////// Constructor(s) //////////
+
+  // Constructor is private to ensure it is a singleton
+  private ProductLoader( Context context )
+    {
+    mContext = context;
+    }
+
+
+  ////////// BaseRequest.BaseRequestListener Method(s) //////////
+
+  /****************************************************
+   *
+   * Called when the retrieval request completes successfully.
+   *
+   ****************************************************/
+  @Override
+  public void onSuccess( int httpStatusCode, JSONObject jsonData )
+    {
+
+    try
+      {
+      // Check if we got a valid HTTP response code
+
+      if ( httpStatusCode >= 200 && httpStatusCode <= 299 )
+        {
+        // Try to getCost an array of products, then parse it.
+
+        JSONArray productsJSONArray = jsonData.getJSONArray( JSON_NAME_PRODUCT_ARRAY );
+
+        Pair<ArrayList<ProductGroup>,HashMap<String,Product>> productPair = parseProducts( productsJSONArray );
+
+        ArrayList<ProductGroup> productGroupList = productPair.first;
+        HashMap<String,Product> productTable     = productPair.second;
+
+        synchronized ( this )
```

Comment:
Your intention may be that `getAllProducts` is safe to call from any thread and the comment is wrong?